File: cloud_testing/cloud_api_poc/src/main/java/cloud_api_poc/Parallel.java
package cloud_api_poc;

import static cloud_api_poc.Utils.fatalError;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class Parallel {
  private List<Callable<Object>> callables;
  private ExecutorService executorService;

  public Parallel(int shardCount) {
    callables = new ArrayList<>(shardCount);
    executorService = Executors.newFixedThreadPool(shardCount);
  }

  public void addCallable(Callable callable) {
    callables.add(callable);
  }

  public void run() {
    try {
      executorService
          .invokeAll(callables)
          .stream()
          .map(
              future -> {
                try {
                  return future.get();
                } catch (Exception e) {
                  throw new IllegalStateException(e);
                }
              });
    } catch (InterruptedException e) {
      fatalError(e);
    } finally {
      executorService.shutdown();
    }
  }
}

File: cloud_testing/kotlin_poc/src/main/kotlin/ftl/Parallel.kt
package ftl

import ftl.Utils.fatalError
import java.util.*
import java.util.concurrent.Callable
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class Parallel(shardCount: Int) {
    private val callables: MutableList<Callable<Any>>
    private val executorService: ExecutorService

    init {
        callables = ArrayList(shardCount)
        executorService = Executors.newFixedThreadPool(shardCount)
    }

    fun addCallable(callable: Callable<Any>) {
        callables.add(callable)
    }

    fun run() {
        try {
            executorService
                    .invokeAll(callables)
                    .stream()
                    .map<Any> { future ->
                        future.get()
                    }
        } catch (e: InterruptedException) {
            fatalError(e)
        } finally {
            executorService.shutdown()
        }
    }
}
